src/library_eventloop.js

```diff
@@ -155,6 +155,61 @@ LibraryJSEventLoop = {
     {{{ runtimeKeepalivePop() }}}
     clearInterval(id);
   },
+
+  $promiseMap__deps: ['$handleAllocator'],
+  $promiseMap: "new handleAllocator();",
+
+  // Create a new promise that can be resolved or rejected by passing a unique
+  // ID to emscripten_promise_resolve/emscripten_promise_reject.  Returns a JS
+  // object containing the promise, its unique ID, and the associated
+  // reject/resolve functions.
+  $newNativePromise__deps: ['$promiseMap'],
+  $newNativePromise: function(funcPtr, userData) {
+    var nativePromise = {};
+    var promiseId;
+    var promise = new Promise((resolve, reject) => {
+      nativePromise.reject = reject;
+      nativePromise.resolve = resolve;
+      nativePromise.id = promiseMap.allocate(nativePromise);
+#if RUNTIME_DEBUG
+      dbg('newNativePromise: ' + nativePromise.id);
+#endif
+      {{{ makeDynCall('vpi', 'funcPtr') }}}(userData, nativePromise.id);
+    });
+    nativePromise.promise = promise;
+    return nativePromise;
+  },
+
+  $getPromise__deps: ['$promiseMap'],
+  $getPromise: function(id) {
+    return promiseMap.get(id).promise;
+  },
+
+  emscripten_promise_create__deps: ['$newNativePromise'],
+  emscripten_promise_create: function(funcPtr, userData) {
+    return newNativePromise(funcPtr, userData).id;
+  },
+
+  emscripten_promise_resolve__deps: ['$promiseMap'],
+  emscripten_promise_resolve__sig: 'vip',
+  emscripten_promise_resolve: function(id, value) {
+#if RUNTIME_DEBUG
+    err('emscripten_resolve_promise: ' + id);
+#endif
+    promiseMap.get(id).resolve(value);
+    promiseMap.free(id);
```

Comment:
How about freeing the id first to get more reuse in case the resolve or reject functions want to allocate more promises?

Reply:
I'm not sure it worth extra line of code (since it would need to stack the promise in a local)... I don't think promise ID reuse is something need to prioritize (do you?) .   In fact, shouldn't we try to avoid it where possible to ease debugging.